### PR TITLE
auth service setup

### LINKED ISSUE #6: Authentication service implementation

### PR DESCRIPTION
**Pull request by mlakdawala-sf**

### CODE REVIEW

File: libs/core/pom.xml [hunk 1]
```diff
@@ -6,7 +6,7 @@
 		<groupId>com.sourcefuse</groupId>
 	    <artifactId>jarc</artifactId>
 	    <version>0.0.1-SNAPSHOT</version>
-        <relativePath>../../</relativePath>
+        <relativePath>../../pom.xml</relativePath>
```
Comment (samarpan-b) [inline libs/core/pom.xml:9]: why is this needed ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/Constants.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.sourcefuse.jarc.services.authservice;
+
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+public class Constants {
+
+  public static final String CURRENT_USER_KEY = "current-user";
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/Constants.java]: Move this to core

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/MyController.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.sourcefuse.jarc.services.authservice.controller;
+
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api")
+public class MyController {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/MyController.java]: why this ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.sourcefuse.jarc.services.authservice.controller;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/")
+public class PingController {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java]: remove this and use https://reflectoring.io/spring-boot-health-check/

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthErrorKeys.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum AuthErrorKeys {
+  CODE_EXPIRED("Code Expired"),
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthErrorKeys.java]: Should we not have them as numeric enums ? We will anyways need to use i18n for localisation

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthProvider.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum AuthProvider {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthProvider.java:3]: Numeric enums ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum RoleKey {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java]: no need of putting in numbers here

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum TenantStatus {
+  ACTIVE(1),
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java]: no need of putting in numbers

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum UserStatus {
+  REGISTERED(0),
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java]: no need of numbers

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum TenantStatus {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java]: shud move to core

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java (hunk 7 above)
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java]: should move to core

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/Gender.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum Gender {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/Gender.java]: should move to core

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package com.sourcefuse.jarc.services.authservice.enums;
+
+public enum UserStatus {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java]: shud move to core

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/CommonRuntimeException.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package com.sourcefuse.jarc.services.authservice.exception;
+
+import org.springframework.http.HttpStatus;
+
+public class CommonRuntimeException extends RuntimeException {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/CommonRuntimeException.java]: why is this needed ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/GlobalExceptionHandler.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.validation.FieldError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/GlobalExceptionHandler.java]: Move to core. @harshadk-sourcefuse you also did similar stuff. Please discuss and do it common.

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/ResourceNotFoundException.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
+package com.sourcefuse.jarc.services.authservice.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class ResourceNotFoundException extends RuntimeException {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/ResourceNotFoundException.java]: WHY ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class BaseEntity extends SoftDeleteEntity {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java]: remove these and rebase on top of @harshadk-sourcefuse PR

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/payload/ClientDTO.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package com.sourcefuse.jarc.services.authservice.payload;
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/payload/ClientDTO.java]: name this package dtos

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/providers/JwtPayloadProvider.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+import com.sourcefuse.jarc.services.authservice.models.User;
+import com.sourcefuse.jarc.services.authservice.models.UserTenant;
+import com.sourcefuse.jarc.services.authservice.session.CurrentUser;
+import org.springframework.stereotype.Service;
+
+@Service
+public class JwtPayloadProvider {
+
+  public CurrentUser provide(User user, UserTenant userTenant, Role role) {
+    return new CurrentUser(user, userTenant, role);
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/providers/JwtPayloadProvider.java]: why no BL here ?

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/services/KeycloakFacadeService.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class KeycloakFacadeService {
```
Comment (samarpan-b) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/services/KeycloakFacadeService.java:20]: Whats this service meant for ?

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: libs/core/pom.xml (hunk 1 above)
Comment (mlakdawala-sf) [reply libs/core/pom.xml:9]: it needed the root pom file to match else it was throwing error in intellij

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/MyController.java (hunk 3 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/MyController.java]: That was for testing auth protected routes will remove it

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java (hunk 4 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java]: will check

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java (hunk 9 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java]: I guess we store numbers in DB hence to map on the basis for the same

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/services/KeycloakFacadeService.java (hunk 19 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/services/KeycloakFacadeService.java:20]: this service has 2 functions, one is for for fetching the user profile from keycloak using token, 2nd one is for fetching the token using the code provided

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java (hunk 16 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java]: Waiting for his PR to be merged

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/CommonRuntimeException.java (hunk 13 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/exception/CommonRuntimeException.java]: image
Just sending prettified JSON Response to the user, basically copied from somewhere and liked it, let me know if we wanna keep this

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java (hunk 7 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java]: We store numbers in DB hence to map that value here we need numbers

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/AuthController.java [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+@RequestMapping("/auth")
+public class AuthController {
+
+  private final AuthService authService;
+  private final JwtService jwtService;
+  private final ClientPasswordVerifyProvider clientPasswordVerifyProvider;
+  private final ResourceOwnerVerifyProvider resourceOwnerVerifyProvider;
+
+  @PostMapping(value = { "/login" })
+  public ResponseEntity<CodeResponse> login(@RequestBody LoginDto loginDto) {
```
Comment (harshadk-sourcefuse) [inline services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/AuthController.java]: does requestBody allowed to have null field?
 if not then use code like below it will validate the requestBody and will throw response with HTTP status badRequest.
 
 ```
 import jakarta.validation.Valid;
 ...
   public ResponseEntity<CodeResponse> login(@Valid @RequestBody LoginDto loginDto) {
 ```
 
 and Update LoginDto with following code
 
 ```
import jakarta.validation.constraints.NotNull;
....
 public class LoginDto {

  @NotNull
  private String username;
  @NotNull
  private String password;
  @NotNull
  private String clientId;
  @NotNull
  private String clientSecret;
}
 ```

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthProvider.java (hunk 6 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/AuthProvider.java:3]: Done

File: libs/core/pom.xml (hunk 1 above)
Comment (mlakdawala-sf) [reply libs/core/pom.xml:9]: reverted

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/Constants.java (hunk 2 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/Constants.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java (hunk 8 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java (hunk 9 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java (hunk 10 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/TenantStatus.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java (hunk 7 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/Gender.java (hunk 11 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/Gender.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java (hunk 12 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/UserStatus.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java (hunk 16 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/models/base/BaseEntity.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java (hunk 4 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/controller/PingController.java]: done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java (hunk 7 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/enums/RoleKey.java]: Done

File: services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/providers/JwtPayloadProvider.java (hunk 18 above)
Comment (mlakdawala-sf) [reply services/auth-service/src/main/java/com/sourcefuse/jarc/services/authservice/providers/JwtPayloadProvider.java]: Currently the current user instance has 3 objects just passing down the same forward, let me know if i should skim it

File: libs/core/src/main/java/com/sourcefuse/jarc/core/repositories/SoftDeletesRepository.java [hunk 21]
```diff
@@ -46,7 +46,7 @@
   List<T> findAllIncludeSoftDelete(Sort sort);
 
   Page<T> findAllIncludeSoftDelete(Pageable pageable);
-
```
Comment (harshadk-sourcefuse) [inline libs/core/src/main/java/com/sourcefuse/jarc/core/repositories/SoftDeletesRepository.java]: revert this file changes

File: services/audit-service/pom.xml [hunk 22]
```diff
@@ -110,7 +110,7 @@
 	        <artifactId>jarc-core</artifactId>
 	        <version>0.0.1-SNAPSHOT</version>
         </dependency>
-	</dependencies>
```
Comment (harshadk-sourcefuse) [inline services/audit-service/pom.xml]: revert this file changes

**Review by harshadk-sourcefuse — CHANGES_REQUESTED** [review]
Rest looks good

**Review by harshadk-sourcefuse — APPROVED** [review]

**Review by samarpan-b — APPROVED** [review]
Approving it for now. We will handle further issues in next PR
